File: MyPane/MyPaneCSharp/MyPaneControl.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using Manifold.Interop;

namespace MyPaneCSharp
{
    public partial class MyPaneControl : UserControl, Manifold.Interop.Scripts.IEventsConnection
    {
        public static Manifold.Interop.Application ManifoldApplication;

        public MyPaneControl()
        {
            InitializeComponent();
        }

        public void ConnectEvents(Manifold.Interop.Scripts.Events ev)
        {
            ev.DocumentClosed += new Manifold.Interop.Scripts.Events.DocumentClosedEventHandler(DocumentChanged);
            ev.DocumentCreated += new Manifold.Interop.Scripts.Events.DocumentCreatedEventHandler(DocumentChanged);
            ev.DocumentOpened += new Manifold.Interop.Scripts.Events.DocumentOpenedEventHandler(DocumentChanged);
            ev.DocumentSaved += new Manifold.Interop.Scripts.Events.DocumentSavedEventHandler(DocumentChanged);
            ev.AddinLoaded += new Manifold.Interop.Scripts.Events.AddinLoadedEventHandler(DocumentChanged);
            ev.ComponentsRemoved += new Manifold.Interop.Scripts.Events.ComponentsRemovedEventHandler(DocumentChanged);
            ev.ComponentStateChanged += new Manifold.Interop.Scripts.Events.ComponentStateChangedEventHandler(DocumentChanged);
            ev.ComponentNameChanged += new Manifold.Interop.Scripts.Events.ComponentNameChangedEventHandler(DocumentChanged);
            ev.ComponentsRemoved += new Manifold.Interop.Scripts.Events.ComponentsRemovedEventHandler(DocumentChanged);

        }

        private void DocumentChanged(System.Object sender, Manifold.Interop.Scripts.DocumentEventArgs args)
        {
            ManifoldApplication = args.Document.Application;
            Label1.Text = args.Document.Path;
        }
        private void DocumentChanged(System.Object sender, Manifold.Interop.Scripts.ComponentEventArgs args)
        {
            ManifoldApplication = args.Component.Application;
        }

        private void Button1_Click(object sender, System.EventArgs e)
        {
            if (ManifoldApplication != null)
            {
                ManifoldApplication.MessageBox("The document contains " +
                  ManifoldApplication.ActiveDocument.ComponentSet.Count.ToString() +
                  " component(s).");
            }
        }
    }
}

File: MyPane/MyPaneVB/MyPaneControl.vb
Public Class MyPaneControl
    Implements Manifold.Interop.Scripts.IEventsConnection

    Dim ManifoldApplication As Manifold.Interop.Application

    Private Sub Button1_Click(
      ByVal sender As System.Object,
      ByVal e As System.EventArgs) Handles Button1.Click
        If Not (ManifoldApplication Is Nothing) Then
            ManifoldApplication.MessageBox("The document contains " +
              ManifoldApplication.ActiveDocument.ComponentSet.Count.ToString() +
              " component(s).")
        End If
    End Sub

    Public Sub ConnectEvents(
      ByVal ev As Manifold.Interop.Scripts.Events) _
      Implements Manifold.Interop.Scripts.IEventsConnection.ConnectEvents
        AddHandler ev.DocumentClosed, AddressOf Document_Changed
        AddHandler ev.DocumentCreated, AddressOf Document_Changed
        AddHandler ev.DocumentOpened, AddressOf Document_Changed
        AddHandler ev.DocumentSaved, AddressOf Document_Changed
        AddHandler ev.AddinLoaded, AddressOf Document_Changed
        AddHandler ev.ComponentsAdded, AddressOf Document_Changed
        AddHandler ev.ComponentsRemoved, AddressOf Document_Changed
        AddHandler ev.ComponentStateChanged, AddressOf Document_Changed
        AddHandler ev.ComponentNameChanged, AddressOf Document_Changed
        AddHandler ev.ComponentsRemoved, AddressOf Document_Changed


    End Sub

    Private Sub Document_Changed(
      ByVal sender As System.Object,
      ByVal args As Manifold.Interop.Scripts.DocumentEventArgs)
        ManifoldApplication = args.Document.Application
        Label1.Text = ManifoldApplication.ActiveDocument.Path
    End Sub

    Private Sub Document_Changed(
      ByVal sender As System.Object,
      ByVal args As Manifold.Interop.Scripts.ComponentEventArgs)
        ManifoldApplication = args.Component.Application
    End Sub

End Class